src/trusted/service_runtime/nacl_app_thread.c

```diff
@@ -710,6 +812,113 @@ void NaClAppThreadDelete(struct NaClAppThread *natp) {
   NaClAlignedFree(natp);
 }
 
+
+/**
+ * The following functions are used to reap any cages which have received a fatal signal.
+ * On launch, sel_main creates the Reaper thread which calls the fault teardown functions when
+ * a faulted thread is caught in the signal handler (nacl_signal.c)
+ * 
+ * The fault teardown function will cycle through all child threads in cage,
+ * tear them down, and call pthread_cancle upon them
+ * 
+ * Finally it will exit the parent thread and signal that the cage has exited with SIGKILL
+ * 
+ * Note: We implement it this way because
+ * 1. We can't fully teardown threads within the signal handler since
+ *  a. We can't unregister/unmap the handler while within itself
+ *  b. Most libc functions are invalid within the handler
+ * 2. We need to not only teardown the faulting thread itself, but any other threads that have been
+ *    launched wtihin that cage.
+ */
+
+
+void InitFatalThreadTeardown(void) {
+  if (!NaClMutexCtor(&teardown_mutex)) {
+    NaClLog(LOG_FATAL, "%s\n", "Failed to initialize handler cleanup mutex");
+  }
+    if (!NaClMutexCtor(&reapermut)) {
+    NaClLog(LOG_FATAL, "%s\n", "Failed to initialize reaper mutex");
+  }
+    if (!NaClCondVarCtor(&reapercv)) {
+    NaClLog(LOG_FATAL, "%s\n", "Failed to initialize reaper cv");
+  }
+}
+
+void DestroyFatalThreadTeardown(void) {
+  NaClMutexDtor(&teardown_mutex);
+  NaClCondVarDtor(&reapercv);
+  NaClMutexDtor(&reapermut);
+}
+
+
+void AddToFatalThreadTeardown(struct NaClAppThread *natp) {
+    if (natp_to_teardown == natp) return;
+    NaClXMutexLock(&teardown_mutex);
+    natp_to_teardown = natp;
+    natp->tearing_down = true;
+    NaClXCondVarBroadcast(&reapercv);
+    NaClXMutexUnlock(&teardown_mutex);
+}
+
```

Comment:
Hmm how come this doesn't just let anyone clobber over the values in this field whenever they want after they're both assigned?